### ISSUE TITLE
tabular mapping: can't create a map when certain characters are in the header for a column

### ISSUE THREAD

**Issue by raprasad** [IQSS/dataverse#3489]
## Updated description by @pdurbin 2017-04

As of [this writing][], users are unable to create a map based on tabular data with certain column names in the header. Here are example files (remove ".txt" to test) and descriptions of column headers are known to fail:

- starting with a digit such as "123Police District ID"
  - Error: `"Error Creating Join: Error Joining table starts_with to layer geonode:police_districts_2014_boston_a7u: syntax error at or near "police_district_id"\nLINE 1: select count(123police_district_id) from starts_with where s...\n ^\n"`
  - Test file: [starts-with-number.csv.txt](https://github.com/IQSS/dataverse/files/913823/starts-with-number.csv.txt)
- accented characters such as "MARK3QTÉ"
  - Error: `UnicodeDecodeError` in logs, button doesn't give feedback (#3763)
  - Test file: [accent.csv.txt](https://github.com/IQSS/dataverse/files/913845/accent.csv.txt)
- only special characters such as "()" or "!@#$%"
  - Error: `"Sorry! Generate SQL to create table from csv file. Error:"`
  - Test file: [just-parens.csv.txt](https://github.com/IQSS/dataverse/files/913864/just-parens.csv.txt)

![screen shot 2017-04-10 at 10 11 27 am](https://cloud.githubusercontent.com/assets/21006/24865870/98a525a6-1dd6-11e7-9070-67fa6f65075d.png)

![screen shot 2017-04-03 at 9 56 31 am](https://cloud.githubusercontent.com/assets/21006/24612643/db05288e-1853-11e7-9448-61d7a81e0d42.png)

Other failures are possible, of course, but the cases above are the ones we know about.

The pull request is in the Geoconnect code base at https://github.com/IQSS/geoconnect/pull/127

[this writing]: https://github.com/IQSS/geoconnect/commit/98289afc2c4928f8cf435d00229585f129005992 

---

## Previous description by @raprasad 2016-11

This is within the Geoconnect codebase.  

dataframe name check fails when orig column name has parentheses

```
if not self.table_attribute_for_join in df.columns:
```

check if due to earlier formatting or something else....

**Comment by pdurbin** [comment]
I'm not sure how to reproduce this bug. I took the `starbucks_us_locations2.csv` file we've been testing with and changed "Latitude" to "Latitude (lat)" in the header and I was still able to create a map:

![screen shot 2017-04-03 at 9 49 38 am](https://cloud.githubusercontent.com/assets/21006/24612361/18f32746-1853-11e7-8377-1a6b874d6ecb.png)

Here's the file I used: [parens.csv.txt](https://github.com/IQSS/dataverse/files/890233/parens.csv.txt)

I'm probably misunderstanding the bug report.

**Comment by pdurbin** [comment]
Ah ha, I think I was just able to reproduce the bug by changing "Latitude" to "()". Now I'm getting "Sorry! Generate SQL to create table from csv file. Error:"

![screen shot 2017-04-03 at 9 56 31 am](https://cloud.githubusercontent.com/assets/21006/24612643/db05288e-1853-11e7-9448-61d7a81e0d42.png)

Here's the file I used: [parens2.csv.txt](https://github.com/IQSS/dataverse/files/890253/parens2.csv.txt)

Some output the might be helpful for debugging:

`create_map_from_datatable_lat_lng 2`

`map_params {'datafile_create_datetime': datetime.datetime(2017, 4, 3, 13, 53, 55, 892000, tzinfo=<UTC>), 'datafile_label': u'parens2.tab', 'abstract': u'This map was created from the file <b>parens2.tab</b>\n\n\n\n    <p><b>Citation:</b> Admin, Dataverse, 2017, "parens", <a href="http://dx.doi.org/10.5072/FK2/H1TAZF" target="_blank">doi:10.5072/FK2/H1TAZF</a>, Root Dataverse, V2, UNF:6:aSjmYv/jBGc1LBCEI2ACNQ==\n\n\n\n    <p><b>Source:</b> Dataset <b>parens</b> stored in Dataverse <b>Root</b></p>\n\n    \n        <p><a href="http://localhost:8080/file.xhtml?fileId=2753&amp;version=2.0">View Data set in Dataverse</a></p>', 'dataverse_installation_name': u'murphy.local', 'dataset_description': u'', 'dv_user_id': 1, 'return_to_dataverse_url': u'http://localhost:8080/file.xhtml?fileId=2753&version=2.0', 'datafile_content_type': u'text/tab-separated-values', 'dataverse_name': u'Root', 'dataset_id': 2749, 'dataset_is_public': True, 'lat_attribute': u'()', 'datafile_expected_md5_checksum': u'd292f68621ce299360ff1b5076b9683f87afdba2', 'dataverse_id': 1, 'datafile_download_url': u'http://localhost:8080/api/access/datafile/2753', 'title': u'parens2.tab', 'dv_user_email': u'dataverse@mailinator.com', 'dataset_semantic_version': u'2.0', 'dataset_citation': u'Admin, Dataverse, 2017, "parens", <a href="http://dx.doi.org/10.5072/FK2/H1TAZF" target="_blank">doi:10.5072/FK2/H1TAZF</a>, Root Dataverse, V2, UNF:6:aSjmYv/jBGc1LBCEI2ACNQ==', 'datafile_filesize': 934495, 'dataverse_description': u'The root dataverse.', 'datafile_id': 2753, 'delimiter': u'\t', 'lng_attribute': u'Longitude', 'dv_username': u'dataverseAdmin', 'dataset_name': u'parens', 'dataset_version_id': 857}`

`create_map_from_datatable_lat_lng 3`

`make request to http://worldmap.harvard.edu/dataverse/api/tabular/upload-lat-lng/`

`----------------------------------------`

`{u'message': u"Generate SQL to create table from csv file.  Error: <class 'sqlalchemy.exc.ArgumentError'>", u'success': False}`

**Comment by pdurbin** [comment]
@raprasad and I talked about this and agreed that we should swarm on issues already in the sprint (already in flight) so I'm dragging this back to the "Next Sprint (4/5-4/19)" column at https://waffle.io/IQSS/dataverse

Before we pick this up again we need a little more information on if the bug I found this morning about "Sorry! Generate SQL to create table from csv file. Error:" is the bug in question or if there are additional bugs that in scope. "Definition of done" needs to be defined, basically.

**Comment by raprasad** [comment]
(all in the geoconnect codebase)

Need some testing for: ```TableJoinMapMaker```

Scenario: Use attempting to map a tabular file (not by lat/lng)

1.  User views column names from dropdown to use for the join
     - These columns names are from:
     - file: ```gc_apps/gis_tabular/views.py```
     - line 366: ```                    column_names=tab_file_stats.column_names)```
       - e.g. scroll up to see initalization of a ```TabFileStats``` object
1.  User submits form, with chose column name
     - file: ```gc_apps/worldmap_connect/table_join_map_maker.py```
     - Geoconnect opens the tabular file via pandas to check if the column selected in (1) is indeed a column name
1.  Info is sent to WorldMap which standardizes all column names as follows:
     - This is **within** WorldMap
 
````python
THE_GEOM_LAYER_COLUMN = 'the_geom'
THE_GEOM_LAYER_COLUMN_REPLACEMENT = 'the_geom_col'


def standardize_column_name(col_name):
    """
    Format column names in tabular files
     - Handle special case for columns named "the_geom", replace them
    """
    assert col_name is not None, "col_name cannot be None"

    cname = slugify(unicode(col_name)).replace('-', '_')
    if cname == THE_GEOM_LAYER_COLUMN:
        return THE_GEOM_LAYER_COLUMN_REPLACEMENT
    return cname

---

There are cases where ```slugify(unicode(col_name)).replace('-', '_')``` will lead to zero length chars or otherwise.

Solution may be to:
- Use same column standardization **before** sending info to WorldMap
  - Set names that fail standardization to something like: ```column_001, column_002, etc.```
- Make sure all column names are unique

**Comment by pdurbin** [comment]
Here's the `standardize_column_name` method: https://github.com/cga-harvard/cga-worldmap/blob/87eec4322c5e0333cd8b7d41149555558bb2a172/geonode/contrib/datatables/name_helper.py#L13

**Comment by pdurbin** [comment]
One of the failing cases is when the column name starts with a number. If you rename rename "Police District ID" to "123Police District ID" as in the starts-with-number.csv test file below (remove ".txt", based on election_precincts3.csv), you should expect the following error from the "master" branch of Geoconnect:

![screen shot 2017-04-10 at 10 11 27 am](https://cloud.githubusercontent.com/assets/21006/24865870/98a525a6-1dd6-11e7-9070-67fa6f65075d.png)

Here's what appears in the log:

```
make request to http://worldmap.harvard.edu/dataverse/api/tabular/upload-join/
----------------------------------------
```

rjson: {
    "message": "Error Creating Join: Error Joining table starts_with to layer geonode:police_districts_2014_boston_a7u: syntax error at or near \"police_district_id\"\nLINE 1: select count(123police_district_id) from starts_with where s...\n                        ^\n", 
    "success": false
}
success: False
error msg: {"message": "Sorry! Error Creating Join: Error Joining table starts_with to layer geonode:police_districts_2014_boston_a7u: syntax error at or near \"police_district_id\"\nLINE 1: select count(123police_district_id) from starts_with where s...\n                        ^\n", "success": false}

[starts-with-number.csv.txt](https://github.com/IQSS/dataverse/files/910328/starts-with-number.csv.txt)

Next I'll be trying to get Geoconnect to use the new `normalize_colname` method I added: https://github.com/IQSS/geoconnect/blob/b5159526953100628ad418f4b0c97cb723736de3/gc_apps/geo_utils/tabular_util.py#L41

**Comment by pdurbin** [comment]
In https://github.com/IQSS/geoconnect/commit/8f3f47ece032e54339e44d1716803134804351b7 I pushed an idea to rename the column names in the CSV/TSV file before they are shown to the user. It doesn't look as nice with the SQL-safe names but at least you can create a map:

## Before: UI looks nice but error creating map

![screen shot 2017-04-10 at 10 11 27 am](https://cloud.githubusercontent.com/assets/21006/24870764/5e19e840-1de5-11e7-8e15-fb106b7753ec.png)

## After: UI looks a bit more robotic but you can create a map

![screen shot 2017-04-10 at 11 50 26 am](https://cloud.githubusercontent.com/assets/21006/24870775/67a6cfc2-1de5-11e7-9ba0-8081101bdf6e.png)

There's a bug though... if I re-map a file, a weird `unnamed_0` column appears, like this:

## Bug: `unnamed_0` column

![screen shot 2017-04-10 at 11 57 17 am](https://cloud.githubusercontent.com/assets/21006/24870821/8fcb8772-1de5-11e7-8fe1-a55ec75bdfd0.png)

@raprasad let's meet to decide what the proper approach should be.

**Comment by pdurbin** [comment]
I just made pull request https://github.com/IQSS/geoconnect/pull/127 and am moving this issue to Code Review at https://waffle.io/IQSS/dataverse

@raprasad heads up that I took a quick swing at the `orginal_column_names` idea but abandoned it since it looked trickier than I thought. I think it's fine to show the renamed, "safe" columns in the UI. I also tested again with just parens in a latitude file and our fix works in that case too. I did make the autogenerated columns (i.e. column_007) friendlier, non-zero indexed in https://github.com/IQSS/geoconnect/commit/95f946516e2f333a3542c664235b81817ff93007 and I tested a file with an accent in the column, which lead me to fix a UnicodeDecodeError in https://github.com/IQSS/geoconnect/commit/ac01f09cb41e3f1404405521a762aef98d4e99db

**Comment by pdurbin** [comment]
@raprasad here's the file I'm testing with that has an accent: 

[accent.csv.txt](https://github.com/IQSS/dataverse/files/911368/accent.csv.txt)

**Comment by pdurbin** [comment]
@raprasad and I aren't so sure about the unicode fix in https://github.com/IQSS/geoconnect/commit/ac01f09cb41e3f1404405521a762aef98d4e99db and might want to try something like `unicode_str = unicode(colname, 'utf-8') instead. Also, calls to `get_worldmap_colname_format` should probably be replaced with calls to the new `normalize_colname` method, which might involve some refactoring.

**Comment by pdurbin** [comment]
I reverted https://github.com/IQSS/geoconnect/commit/ac01f09cb41e3f1404405521a762aef98d4e99db so I could see the error ("UnicodeDecodeError ... 'ascii' codec can't decode byte 0xc3 in position 7: ordinal not in range(128)") when uploading a file with an accent. I also created a new issue at #3763 in case we decide this is out of scope. This issue would probably benefit from a write up of "definition of done", especially before it goes to QA. Here's the stack trace:

![screen shot 2017-04-11 at 9 50 13 am](https://cloud.githubusercontent.com/assets/21006/24912583/8ff13596-1e9c-11e7-8d33-e91ef7d602a4.png)

I'm playing around with removing the call to `unicode` entirely. We'll see.

**Comment by pdurbin** [comment]
@raprasad please check out the branch as of https://github.com/IQSS/geoconnect/commit/2eb23122a3fdb4480b4b3d8fdc527451c802e569

I removed calls to `unicode` and I **think** it fixes everything. I'm moving this to Code Review at https://waffle.io/IQSS/dataverse but please feel free to give it back to me if there's more work you want me to do. Please note that I'm not addressing removing `get_worldmap_colname_format` because it isn't much technical debt and it was non-obvious to me how to remove it. I'll probably just close #3763 since I ended up fixing the accents/unicode stuff in this issue after all.

**Comment by raprasad** [comment]
@pdurbin : I'll take over the branch.  ```get_worldmap_colname_format``` needs to work the same way as the new method.

**Comment by raprasad** [comment]
can explain in person with a diagram

**Comment by pdurbin** [comment]
@raprasad and I discussed this issue and we're hoping we can simply delete the `get_worldmap_colname_format` method since the columns have already been normalized at this point. I took a swing at this yesterday but failed so @raprasad is going to try next.

**Comment by pdurbin** [comment]
I just tried again to delete the `get_worldmap_colname_format` method and deleting it didn't cause me any trouble so I pushed the change at cfa5a64 .

I'm hoping the branch is ready. I moved it to Code Review so @raprasad can give it a final blessing or make whatever tweaks he sees fit.

**Comment by raprasad** [comment]
One existing tests now fails.  Will rewrite the test to reflect the new column formatting.

**Comment by raprasad** [comment]
test updated

**Comment by raprasad** [comment]
Back to dev.  Was not tested with S3 for rewriting file with new column headings

**Comment by raprasad** [comment]
working for S3--incorrect command to write file in orig. code

**Comment by raprasad** [comment]
## failing scenario

- map tabular file via join (e.g. zip code) where **not all rows match**
- return to Dataverse
- Geoconnect admin: delete file and metadata (this happens automatically every 24 hours)
- remap file from Dataverse
- Fails: 
   - Geoconnect has retrieved a fresh copy of the tabular file from geoconnect
   - Geoconnect finds existing layer in WorldMap but 
   - Geoconnect DOES NOT rename the columns of the freshly retrieved tabular file.
   - The attempt to display failed rows throws an error b/c of the column name difference
       - Note: geoconnect only knows the join column vals of the failed rows and must pull them from the fresh file

**Comment by raprasad** [comment]
Last failure fixed.  However new bug found.  Column renaming loop within ```normalize_colname``` is comparing unicode to string.

**Comment by raprasad** [comment]
needs some additional code review...perhaps from @pameyer

**Comment by pdurbin** [comment]
I'm slightly confused. I see that https://github.com/IQSS/geoconnect/pull/127 was merged already. Great. That's the bit I tried to help with. Now there's https://github.com/IQSS/geoconnect/pull/130 which seems to have a bit more scope that I expect. I'll grab @raprasad and talk it out.

**Comment by raprasad** [comment]
w/ @pdurbin: merged and deployed PR IQSS/geoconnect#130 to geoconnect test 

@kcondon : this can be tested.  let me know before you do--e.g. to clear out worldmap test files

**Comment by kcondon** [comment]
Looks good.
